fleetspeak/src/client/daemonservice/execution/execution.go

```diff
@@ -378,51 +368,76 @@ func (e *Execution) inLoop() {
 		e.Shutdown()
 		e.inProcess.Done()
 	}()
+	initialMsg := e.readMsg()
+	if initialMsg == nil {
+		return
+	}
+
+	if initialMsg.Destination != nil && initialMsg.Destination.ServiceName == "system" && initialMsg.MessageType == "StartupData" {
+		sd := &fcpb.StartupData{}
+		if err := ptypes.UnmarshalAny(initialMsg.Data, sd); err != nil {
+			log.Warningf("Failed to parse startup data from initial message: %v", err)
+		} else {
+			e.startupData <- sd
+			close(e.startupData) // No more values to send through the channel.
+		}
+	} else {
+		// Handle initial message like we would any other.
+		e.setLastActive(time.Now())
+		if err := e.sc.Send(context.Background(), service.AckMessage{M: initialMsg}); err != nil {
+			log.Errorf("error sending message to server: %v", err)
+		}
+	}
 	for {
-		select {
-		case m, ok := <-e.channel.In:
-			if !ok {
-				return
-			}
-			e.setLastActive(time.Now())
-			if err := e.sc.Send(context.Background(), service.AckMessage{M: m}); err != nil {
-				log.Errorf("error sending message to server: %v", err)
-			}
-		case err := <-e.channel.Err:
-			log.Errorf("channel produced error: %v", err)
+		m := e.readMsg()
+		if m == nil {
 			return
 		}
+		e.setLastActive(time.Now())
+		if err := e.sc.Send(context.Background(), service.AckMessage{M: m}); err != nil {
+			log.Errorf("error sending message to server: %v", err)
+		}
 	}
 }
 
-// sendStats attempts to send an mpb.ResourceUsageData proto to the server.
-func (e *Execution) sendStats(ctx context.Context, aggRU *mpb.AggregatedResourceUsage) {
-	startTime, err := ptypes.TimestampProto(e.StartTime)
-	if err != nil {
-		// Well, this really should never happen, since the field is set to the return
-		// value of time.Now()
-		log.Errorf("Client start time timestamp failed validation check: %v", e.StartTime)
-		return
+// readMsg blocks until a message is available from the channel.
+func (e *Execution) readMsg() *fspb.Message {
+	select {
+	case m, ok := <-e.channel.In:
+		if !ok {
+			return nil
+		}
+		return m
+	case err := <-e.channel.Err:
+		log.Errorf("channel produced error: %v", err)
+		return nil
 	}
-	rud := mpb.ResourceUsageData{
-		Scope:            e.daemonServiceName,
-		Pid:              int64(e.cmd.Process.Pid),
-		ProcessStartTime: startTime,
-		DataTimestamp:    ptypes.TimestampNow(),
-		ResourceUsage:    aggRU,
+}
+
+// statsLoop monitors the daemon process's resource usage, sending reports to the server
+// at regular intervals.
+func (e *Execution) statsLoop() {
+	defer e.inProcess.Done()
+	pid := e.cmd.Process.Pid
+	select {
+	case sd := <-e.startupData:
+		if int(sd.Pid) != pid {
+			log.Infof("%s's self-reported PID (%d) is different from that of the process launched by Fleetspeak (%d)", e.daemonServiceName, sd.Pid, pid)
+			pid = int(sd.Pid)
+		}
+	case <-time.After(startupDataTimeout):
+		log.Errorf("%s failed to send startup data after %v", e.daemonServiceName, startupDataTimeout)
```

Comment:
nit: Warningf and/or add reassurance that it isn't too broken, e.g. "failed to receive startup data after %v, using observed PID."

Reply:
Done.